.github/workflows/pytest.yml

```diff
@@ -0,0 +1,67 @@
+name: Pytest
+on: [push, pull_request, workflow_dispatch]
+
+env:
+  HUGEPAGES: 512
+
+jobs:
+  pytest:
+    strategy:
+      matrix:
+        test: ["cli", "state", "multi-gateway"]
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
+        with:
+          # git submodule update --init --recursive
+          submodules: recursive
+
+      - name: Build images
+        run: |
+          SVC="ceph nvmeof-pytest-${{ matrix.test }}" make build
```

Comment:
This step is common to this workflow and to the 3 matrix runs (cli, state, multi-gateway), so what about merging everything in a single workflow, having the build step as a separate job, pushing containers to the Github registry (tagged as `${{github.run_number}}` or `$GITHUB_RUN_NUMBER`), and consuming those in the follow-up jobs (integration test, unit tests: cli, state, multi-gateway).

That said, I don't think that unit tests should require a real deployment with containers and so, as this testing will be also affected by integration issues, so they might mask whether the component specific changes are ok or not. IMHO the unit testing should rely on mocking and be completely isolated from other components (`cli` should just test that the `cli` Python code is ok, `state` should test against a mocked `nvmeof` RADOS object and a similar thing for `multi-gateway`).

Reply:
> While I agree extracting build images is a good idea overall, I'm not overly concerned about those duplicated steps in several workflows. Pytests typically take around 3 minutes to complete, including the image build, and run in concurrently, so it's not a significant concern for me.

I'm surprised to see that it just takes between 2-3 mins (it goes up to 9 mins in the `build-and-test` job). Ah, ok, I see where it's the trick: the `spdk` image is not built locally, but downloaded:
```sh
#8 [base-gateway 1/2] FROM quay.io/ceph/spdk:23.01@sha256:1f808ef5fcc1c8443bdcfed264a6393187555a58bf86d4ee0cd68e7408c58d35
```

While that's not going to make a big difference most of the time, it won't test changes coming from the `spdk` submodule... I wouldn't mind that much if these were _pure_ unit tests, but as they are integration tests, I think this could be a misleading behaviour. That said, I'm not opposed to get this merged as it is (as it brings value anyway), and fix it in a follow-up PR.

Different question is about the overhead: it takes 3 minutes to build the images and just 2-10 seconds to run the tests. That's a 95-99% overhead 😞 As a comparison, if we were able to perform true (mocked) unit-tests, that'd be an almost "push commit and get feedback".

Even preferring this as integration tests, IMHO [multi-job workflows](https://github.blog/2021-11-18-7-advanced-workflow-automation-features-with-github-actions/#1-track-your-workflows-with-workflow-visualization) would provide a richer (and more efficient) CI feedback.

> Another thought, noticed that the ceph iSCSI team heavily uses jenkins.ceph.com. Does it make sense building images there? My suggestion is to extract the image builds as a separate pull request and concentrate on incorporating the current pytest suite for now.

From experience, I would stay away from Ceph Jenkins as much as possible:
* We here have pipelines as code (that's also possible in Jenkins, but the current approach relies on having Jenkins jobs stored in a separate, repo [ceph/ceph-build](https://github.com/ceph/ceph-build), which may cause inconsistencies issues between pipeline code and tested code)
* It's not uncommon to face environment issues. With GH Actions, we're getting stable/free/maintained CI here.
* IMHO the overall developer experience/integration is much better achieved with Github Actions than with Jenkins.